src/Symfony/Component/Security/Http/Attribute/IsGranted.php

```diff
@@ -24,18 +26,37 @@
 final class IsGranted
 {
     /**
-     * @param string|Expression            $attribute     The attribute that will be checked against a given authentication token and optional subject
-     * @param array|string|Expression|null $subject       An optional subject - e.g. the current object being voted on
-     * @param string|null                  $message       A custom message when access is not granted
-     * @param int|null                     $statusCode    If set, will throw HttpKernel's HttpException with the given $statusCode; if null, Security\Core's AccessDeniedException will be used
-     * @param int|null                     $exceptionCode If set, will add the exception code to thrown exception
+     * @param string|Expression|null                                                           $attribute     The attribute that will be checked against a given authentication token and optional subject, or a callable that will be called to determine access
+     * @param array|string|Expression|\Closure(Request $request, array $arguments): mixed|null $subject       An optional subject - e.g. the current object being voted on
+     * @param string|null                                                                      $message       A custom message when access is not granted
+     * @param int|null                                                                         $statusCode    If set, will throw HttpKernel's HttpException with the given $statusCode; if null, Security\Core's AccessDeniedException will be used
+     * @param int|null                                                                         $exceptionCode If set, will add the exception code to thrown exception
+     * @param \Closure(AuthorizationCheckerInterface, mixed $subject): bool|null               $callable      A callable that will be called to determine access
      */
     public function __construct(
-        public string|Expression $attribute,
-        public array|string|Expression|null $subject = null,
+        public string|Expression|null $attribute = null,
+        public array|string|Expression|\Closure|null $subject = null,
         public ?string $message = null,
         public ?int $statusCode = null,
         public ?int $exceptionCode = null,
+        public ?\Closure $callable = null,
     ) {
+        if (\PHP_VERSION_ID < 80500) {
+            if (null === $attribute) {
+                throw new \LogicException('The "attribute" argument is only optional starting from PHP 8.5.');
+            }
+
+            if (null !== $callable) {
+                throw new \LogicException('The "callable" argument is usable only starting from PHP 8.5.');
+            }
+
+            if ($subject instanceof \Closure) {
+                throw new \LogicException('The "subject" argument can be a callable only starting from PHP 8.5.');
```

Comment:
This exception does not make sense. Our own code has no incompatibility with passing a Closure there. The incompatibility is the PHP syntax for attributes, which will fail **before** calling the constructor.